src/routes/[id]/training/queues/[queueId]/+page.server.ts

```diff
@@ -19,5 +25,60 @@ export const load: PageServerLoad = async ({ params, depends }) => {
 
     return {
         queue,
+        form: await superValidate(formSchema),
     };
+}
+
+export const actions: Actions = {
+    addStudent: async (event) => {
+        const form = await superValidate(event, formSchema);
+
+        if (!can(MANAGE_QUEUES)) {
+            redirect(307, event.url.pathname, {
+                type: "error",
+                message: "You do not have permission to do that"
+            }, event.cookies);
+        }
+
+        const targetUser = await prisma.user.findUnique({
+            where: {
+                id: form.data.id,
+            }
+        });
+
+        if (!targetUser) {
+            redirect(307, event.url.pathname, {
+                type: "error",
+                message: "This user is not registered on Hayya"
+            }, event.cookies);
+        }
+
+        const queueMembership = await prisma.trainingQueueMembership.findMany({
+            where: {
+                queueId: event.params.queueId,
+                userId: form.data.id,
+            }
+        });
+
+        if (queueMembership.length > 0) {
+            redirect(307, event.url.pathname, {
+                type: "error",
+                message: "Student is already a member of this queue"
+            }, event.cookies);
+        }
+
+        if (event.params.queueId) {
+            await prisma.trainingQueueMembership.create({
```

Comment:
This is vulnerable to a race condition. The correct way to handle this would be a unique constraint in the database. If the create fails, that would probably mean that they are already in the queue. Also saves you from doing multiple queries.

Reply:
aed5401095e9a062a556ae8c1b65c65b71bfd377 042b48da39dc7d8da885b0214ebf84fac1659433

I'm not terribly familiar with Prisma. Could you take a look and see if I've done this correctly?